vllm/v1/stats/common.py

```diff
@@ -0,0 +1,393 @@
+        if update.type == RequestStatsUpdate.Type.ARRIVED:
+            self.arrival_ts_s = ts
+        elif update.type == RequestStatsUpdate.Type.INPUT_PROCESSED:
+            self.input_processor_end_ts_s = ts
+            self.engine_request = update.engine_request
+        elif update.type == RequestStatsUpdate.Type.QUEUED:
+            self.queued_ts_s = ts
+        elif update.type == RequestStatsUpdate.Type.RUNNING:
+            assert (update.new_prefill is not None
+                    and update.num_computed_tokens is not None)
+            self._record_running(
+                update.num_computed_tokens,
+                update.new_prefill,
+                ts,
+                update.num_cached_tokens,
+            )
+        elif update.type == RequestStatsUpdate.Type.PREEMPTED:
+            self._reset_for_preemption(ts)
+        elif update.type == RequestStatsUpdate.Type.DECODED:
+            assert update.token_perf_ts_ns is not None
+            self._record_engine_output(
+                ts,
+                update.token_perf_ts_ns,
+                update.num_new_tokens,
+                update.finish_reason,
+            )
+        elif update.type == RequestStatsUpdate.Type.DETOKENIZED:
+            self._record_request_output(update.finish_reason, ts)
+        else:
+            raise ValueError(f"Unknown update type: {update.type}")
+
+    def _record_running(
+        self,
+        num_computed_tokens: int,
+        new_prefill: bool,
+        ts_s: float,
+        num_cached_tokens: Optional[int] = None,
+    ):
+        if new_prefill:
+            # Was preempted or a newly scheduled request - record the
+            # prefill start timestamp.
+            self.prefill_start_ts_s_lst.append(ts_s)
+            self.num_cached_tokens = num_cached_tokens
+
+        self.num_computed_tokens = num_computed_tokens
+
+    def _record_engine_output(
+        self,
+        ts_s: float,
+        perf_ts_ns: int,
+        num_new_tokens: int,
+        finish_reason: Optional[str],
+    ):
+        # Update if first output token is generated.
+        if len(self.output_token_perf_counter_ns_lst) == 0:
+            self.first_token_ts_s = ts_s
+            assert (
+                self.prefill_ts_s is not None
+            ), "Request must be running before generating output tokens."
+
+        self.output_token_perf_counter_ns_lst.extend([perf_ts_ns] *
+                                                     num_new_tokens)
+
+        # Update if the request is finished.
+        if finish_reason is not None:
+            self.finished_ts_s = ts_s
+            self.finish_reason = finish_reason
+
+    def _record_request_output(self, finish_reason: Optional[str],
+                               ts_s: float):
+        if finish_reason is not None and self.finished_ts_s is None:
+            self.finished_ts_s = ts_s
+            self.finish_reason = finish_reason
+
+    def _reset_for_preemption(self, ts_s: float):
```

Comment:
Is resetting the right strategy? E.g. for the `forward_duration_s` and `model_execute_duration_s`?

When we compute higher level metrics based on this, won't resetting underestimate how long is actually spent in forward passes?